src/mac.cpp

```diff
@@ -0,0 +1,112 @@
+/*
+ * #%L
+ * %%
+ * Copyright (C) 2020 BMW Car IT GmbH
+ * %%
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ * #L%
+ */
+#include "mococrw/mac.h"
+#include "mococrw/error.h"
+#include "mococrw/openssl_wrap.h"
+
+#include <openssl/hmac.h>
+
+namespace mococrw
+{
+
+class HMAC::Impl
+{
+public:
+    Impl(openssl::DigestTypes hashFunction, const std::vector<uint8_t> key, size_t outputLength) :
+        outputLength(outputLength)
+    {
+        const EVP_MD* digestFn = _getMDPtrFromDigestType(hashFunction);
+
+        if (key.empty()) {
+            throw MoCOCrWException("Key for HMAC is empty.");
+        }
+
+        ctx = openssl::_HMAC_CTX_new();
```

Comment:
I don't have a good overview of the advantages and disadvantages, but have you considered using the `EVP_*` digest interface to do this? See https://wiki.openssl.org/index.php/EVP_Message_Digests and https://wiki.openssl.org/index.php/EVP_Signing_and_Verifying#Calculating_HMAC for more information.

Reply:
I haven't considered it.
I checked the openssl tests for hmac (test/hmactest.c) and there the HMAC_CTX_new and HMAC_Init_ex interface is used. So I decided to use this, too.

Reply:
I checked it again.
Internally the HMAC_* interface is invoking the EVP_* interface. So this is a "higher" level interface for calculating the MAC.